File: sdks/node/examples/basic.ts
/**
 * Minimal example: init obs-unified at process startup, then use
 * `withLLMSpan` and `withToolSpan` to instrument an agent loop.
 *
 * Run it (after building):
 *   node --import ./dist/init-shim.js dist/examples/basic.js
 *
 * Auto-instrumentation for HTTP / DB clients comes from the OTel
 * ecosystem — install `@opentelemetry/auto-instrumentations-node` and
 * pass `getNodeAutoInstrumentations()` via `instrumentations` in init.
 */

import { init, setProjectId, withLLMSpan, withToolSpan } from "../src";

const shutdown = init({
	collectorUrl: process.env.OBS_COLLECTOR_URL ?? "http://localhost:8790",
	ingestKey: process.env.OBS_INGEST_KEY ?? "",
	serviceName: "example-agent",
	serviceVersion: "0.1.0",
	environment: "dev",
});

const main = async () => {
	setProjectId("default");

	const answer = await withLLMSpan(
		{ provider: "openai", model: "gpt-4o-mini", maxTokens: 256, turnIndex: 0 },
		async (span) => {
			// Pretend HTTP call — auto-instrumented via OTel's HTTP instrumentation
			// in real apps. The response usage is what we care about here.
			const fakeResponse = {
				choices: [{ message: { content: "It is sunny." } }],
				usage: { prompt_tokens: 42, completion_tokens: 5, total_tokens: 47 },
			};
			span.setUsage({
				inputTokens: fakeResponse.usage.prompt_tokens,
				outputTokens: fakeResponse.usage.completion_tokens,
				totalTokens: fakeResponse.usage.total_tokens,
			});
			span.setFinishReason("stop");
			return fakeResponse.choices[0]?.message.content ?? "";
		},
	);

	const items = await withToolSpan(
		{ name: "list_widgets", args: { color: "blue" } },
		async (span) => {
			const out = ["a", "b", "c"];
			span.setOutcome("ok");
			span.setResultCount(out.length);
			return out;
		},
	);

	console.log({ answer, items });
};

main()
	.catch((err) => {
		console.error(err);
		process.exitCode = 1;
	})
	.finally(async () => {
		await shutdown();
	});

File: sdks/rust/examples/basic.rs
//! Minimal example. Run with:
//!
//!   OBS_COLLECTOR_URL=http://localhost:8790 OBS_INGEST_KEY=obs_xxx \
//!     cargo run --example basic
//!
//! In a real app you'd also wire HTTP / DB instrumentation via
//! `tracing-opentelemetry`, `tower-http::trace`, and per-driver crates.

use std::env;

use obs_unified::{
    init, with_llm_span, with_tool_span, Config, LlmOptions, LlmResult, ToolOptions, ToolResult,
};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let _guard = init(Config {
        collector_url: env::var("OBS_COLLECTOR_URL")
            .unwrap_or_else(|_| "http://localhost:8790".into()),
        ingest_key: env::var("OBS_INGEST_KEY").unwrap_or_default(),
        service_name: "example-agent".into(),
        service_version: Some("0.1.0".into()),
        environment: Some("dev".into()),
        project_id: Some("default".into()),
        ..Default::default()
    })?;

    let answer: String = with_llm_span::<_, std::io::Error, _, _>(
        LlmOptions {
            provider: "openai",
            model: "gpt-4o-mini",
            max_tokens: Some(256),
            turn_index: Some(0),
            ..Default::default()
        },
        || async {
            // Pretend HTTP call. Real apps: tracing-opentelemetry will
            // create a child span for the actual reqwest call.
            Ok(LlmResult::new("It is sunny.".to_string())
                .with_usage(42, 5)
                .with_finish_reason("stop"))
        },
    )
    .await?;

    let items: Vec<String> = with_tool_span::<_, std::io::Error, _, _>(
        ToolOptions {
            name: "list_widgets",
            args: Some(serde_json::json!({ "color": "blue" })),
            ..Default::default()
        },
        || async {
            let out = vec!["a".to_string(), "b".to_string(), "c".to_string()];
            let count = out.len() as i64;
            Ok(ToolResult::new(out)
                .with_outcome("ok")
                .with_result_count(count))
        },
    )
    .await?;

    println!("answer={answer:?} items={items:?}");
    Ok(())
}
